src/libcore/hash/sip.rs

```diff
@@ -83,38 +85,40 @@ struct State {
 // this design could be improved.
 
 macro_rules! u8to64_le {
-    ($buf:expr, $i:expr) =>
-    ($buf[0+$i] as u64 |
-     ($buf[1+$i] as u64) << 8 |
-     ($buf[2+$i] as u64) << 16 |
-     ($buf[3+$i] as u64) << 24 |
-     ($buf[4+$i] as u64) << 32 |
-     ($buf[5+$i] as u64) << 40 |
-     ($buf[6+$i] as u64) << 48 |
-     ($buf[7+$i] as u64) << 56);
     ($buf:expr, $i:expr, $len:expr) =>
     ({
         let mut t = 0;
         let mut out = 0;
-        while t < $len {
-            out |= ($buf[t+$i] as u64) << t*8;
-            t += 1;
+        if $len >= 4 {
+            out = load_int_le!($buf, $i, u32) as u64;
+            t = 4;
+        }
+        if t + 1 < $len {
+            out |= (load_int_le!($buf, t+$i, u16) as u64) << t*8;
+            t += 2;
+        }
+        if t < $len {
```

Comment:
Good points. I think a variable renaming will help here, and also making it inline. I just have to check the code gen.

Counting down from $len isn't really desirable because then we need some additional arithmetic to calculate the shifts.